File: src/synchronizer.cpp
// Experimental node for synchronization and stereo image processing

#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <sensor_msgs/CameraInfo.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

class Synchronizer
{
private:
    // Declare node handler, subscriber and publishers
    ros::NodeHandle nh_;
    message_filters::Subscriber<sensor_msgs::Image> sub_1_, sub_2_;
    message_filters::Subscriber<sensor_msgs::CameraInfo> sub_1_i, sub_2_i;
    ros::Publisher pub_1_, pub_2_, pub_1_i, pub_2_i;
    typedef message_filters::sync_policies::ApproximateTime<sensor_msgs::Image, sensor_msgs::Image, sensor_msgs::CameraInfo, sensor_msgs::CameraInfo> MySyncPolicy;
    typedef message_filters::Synchronizer<MySyncPolicy> Sync;
    boost::shared_ptr<Sync> sync_;
    // Declare camera info and header messages
    sensor_msgs::CameraInfo M_CAMERA_INFO, K_CAMERA_INFO;
    // Global variables
    static const short BAND_WIDTH = 960;
    static const short BAND_HEIGHT = 540;
    static constexpr const char *D_MODEL = "plumb_bob";
    static constexpr const char *FRAME_ID = "stereo_frame";
    static constexpr const char *IME_B = "mono8";

public:
    Synchronizer()
    {
        // Subscribers
        sub_1_.subscribe(nh_, "/band/1/image_raw", 1);
        sub_2_.subscribe(nh_, "/kinect2/qhd/image_color_rect", 1);
        sub_1_i.subscribe(nh_, "/band/camera_info", 1);
        sub_2_i.subscribe(nh_, "/kinect2/qhd/camera_info", 1);
        // Publishers
        pub_1_ = nh_.advertise<sensor_msgs::Image>("/stereo/right/image_raw", 1);
        pub_2_ = nh_.advertise<sensor_msgs::Image>("/stereo/left/image_raw", 1);
        pub_1_i = nh_.advertise<sensor_msgs::CameraInfo>("/stereo/right/camera_info", 1);
        pub_2_i = nh_.advertise<sensor_msgs::CameraInfo>("/stereo/left/camera_info", 1);
        // Synchronizer policy
        sync_.reset(new Sync(MySyncPolicy(10), sub_1_, sub_2_, sub_1_i, sub_2_i));
        sync_->registerCallback(boost::bind(&Synchronizer::callback, this, _1, _2, _3, _4));
        // Display info
        ROS_INFO("Synchronizer node started");
        // Init camera info
        M_CAMERA_INFO.header.seq = 0;
        M_CAMERA_INFO.header.frame_id = FRAME_ID;
        M_CAMERA_INFO.width = BAND_WIDTH;
        M_CAMERA_INFO.height = BAND_HEIGHT;
        M_CAMERA_INFO.distortion_model = D_MODEL;

        K_CAMERA_INFO.header.frame_id = FRAME_ID;
        K_CAMERA_INFO.header.seq = 0;
        K_CAMERA_INFO.width = BAND_WIDTH;
        K_CAMERA_INFO.height = BAND_HEIGHT;
        K_CAMERA_INFO.distortion_model = D_MODEL;
    }

    void callback(const sensor_msgs::ImageConstPtr &in1, const sensor_msgs::ImageConstPtr &in2, const sensor_msgs::CameraInfoConstPtr &in3, const sensor_msgs::CameraInfoConstPtr &in4)
    {
        // Set to the camera info the time stamp
        M_CAMERA_INFO.header.stamp = ros::Time::now();
        K_CAMERA_INFO.header.stamp = ros::Time::now();

        // Convert from ROS to OpenCV
        cv_bridge::CvImagePtr im_ptr1, im_ptr2;
        try
        {
            im_ptr1 = cv_bridge::toCvCopy(in1, sensor_msgs::image_encodings::MONO8);
            im_ptr2 = cv_bridge::toCvCopy(in2, sensor_msgs::image_encodings::MONO8);
        }
        catch (cv_bridge::Exception &e)
        {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }
        // Image message to opencv image
        cv::Mat img_m = im_ptr1->image;
        cv::Mat img_k = im_ptr2->image;

        // Set camera info
        M_CAMERA_INFO.D = in3->D;
        M_CAMERA_INFO.K = in3->K;
        M_CAMERA_INFO.R = in3->R;
        M_CAMERA_INFO.P = in3->P;

        K_CAMERA_INFO.D = in4->D;
        K_CAMERA_INFO.K = in4->K;
        K_CAMERA_INFO.R = in4->R;
        K_CAMERA_INFO.P = in4->P;

        // Resize image
        cv::Mat img_rm;
        resize(img_m, img_rm);

        // Convert to image message
        sensor_msgs::ImagePtr imgCVM = cv_bridge::CvImage(M_CAMERA_INFO.header, IME_B, img_rm).toImageMsg();
        sensor_msgs::ImagePtr imgCVK = cv_bridge::CvImage(K_CAMERA_INFO.header, IME_B, img_k).toImageMsg();

        // Publish images
        pub_1_.publish(imgCVM);
        pub_2_.publish(imgCVK);
        pub_1_i.publish(M_CAMERA_INFO);
        pub_2_i.publish(K_CAMERA_INFO);
    }

    // Resize multispectral image to fit
    void resize(cv::Mat &img_m, cv::Mat &result)
    {
        // Bicubic Interpolation for reseizing
        cv::Mat img_r = cv::Mat(764, 960, CV_8UC1);
        cv::resize(img_m, img_r, img_r.size(), 0, 0, cv::INTER_CUBIC);
        // Crop images to fit with depth images 960x540
        img_r(cv::Rect(cv::Point(0, 652), cv::Point(960, 112))).copyTo(result);
    }

    // Shut down everything
    void stop()
    {
        pub_1_.shutdown();
        pub_2_.shutdown();
        pub_1_i.shutdown();
        pub_2_i.shutdown();
        ros::shutdown();
        std::cout << "Synchronizer node shutted down" << std::endl;
    }
};

int main(int argc, char **argv)
{
    ros::init(argc, argv, "synchronizer");
    Synchronizer synchronizer;
    ros::MultiThreadedSpinner spinner(4);
    spinner.spin();
    synchronizer.stop();
    return 0;
}

File: src/synchronizer.py
#!/usr/bin/env python

from __future__ import print_function
from cv_bridge import CvBridge, CvBridgeError
from sensor_msgs.msg import CameraInfo
from sensor_msgs.msg import Image
import message_filters
import rospy
import cv2
import sys
import time
import roslib
import numpy as np

# Experimental node for synchronization and stereo image processing


class Synchronizer:
    # Global variables
    BAND_WIDTH = 960
    BAND_HEIGHT = 540
    D_MODEL = "plumb_bob"
    FRAME_ID = "stereo_frame"
    IME_B = "mono8"

    def __init__(self):
        # Optimization in OpenCV
        cv2.useOptimized()
        # Initialize cv bridge
        self.bridge = CvBridge()
        # Subcribers
        self.sub_1_ = message_filters.Subscriber("/band/1/image_raw", Image)
        self.sub_2_ = message_filters.Subscriber(
            "/kinect2/qhd/image_color_rect", Image)
        self.sub_1_i = message_filters.Subscriber(
            "/band/camera_info", CameraInfo)
        self.sub_2_i = message_filters.Subscriber(
            "/kinect2/qhd/camera_info", CameraInfo)
        # Publishers
        self.pub_1_ = rospy.Publisher(
            "/stereo/right/image_raw", Image, queue_size=1)
        self.pub_2_ = rospy.Publisher(
            "/stereo/left/image_raw", Image, queue_size=1)
        self.pub_1_i = rospy.Publisher(
            "/stereo/right/camera_info", CameraInfo, queue_size=1)
        self.pub_2_i = rospy.Publisher(
            "/stereo/left/camera_info", CameraInfo, queue_size=1)
        # Synchronizer
        ts = message_filters.ApproximateTimeSynchronizer(
            [self.sub_1_, self.sub_2_, self.sub_1_i, self.sub_2_i], 10, 0.1, allow_headerless=True)
        ts.registerCallback(self.callback)
        # Display info
        rospy.loginfo("Synchronizer node (python) started")
        # Init camera info
        self.M_CAMERA_INFO = CameraInfo()
        self.M_CAMERA_INFO.header.seq = 0
        self.M_CAMERA_INFO.header.frame_id = self.FRAME_ID
        self.M_CAMERA_INFO.width = self.BAND_WIDTH
        self.M_CAMERA_INFO.height = self.BAND_HEIGHT
        self.M_CAMERA_INFO.distortion_model = self.D_MODEL

        self.K_CAMERA_INFO = CameraInfo()
        self.K_CAMERA_INFO.header.frame_id = self.FRAME_ID
        self.K_CAMERA_INFO.header.seq = 0
        self.K_CAMERA_INFO.width = self.BAND_WIDTH
        self.K_CAMERA_INFO.height = self.BAND_HEIGHT
        self.K_CAMERA_INFO.distortion_model = self.D_MODEL

    def callback(self, in1, in2, in3, in4):
        # Set to the camera info the time stamp
        self.M_CAMERA_INFO.header.stamp = rospy.Time.from_sec(time.time())
        self.K_CAMERA_INFO.header.stamp = rospy.Time.from_sec(time.time())

        # Convert from ROS to OpenCV
        try:
            img1 = self.bridge.imgmsg_to_cv2(in1, self.IME_B)
            img2 = self.bridge.imgmsg_to_cv2(in2, self.IME_B)
        except CvBridgeError as e:
            print(e)

        # Set camera info
        self.M_CAMERA_INFO.D = in3.D
        self.M_CAMERA_INFO.K = in3.K
        self.M_CAMERA_INFO.R = in3.R
        self.M_CAMERA_INFO.P = in3.P

        self.K_CAMERA_INFO.D = in4.D
        self.K_CAMERA_INFO.K = in4.K
        self.K_CAMERA_INFO.R = in4.R
        self.K_CAMERA_INFO.P = in4.P

        # Resize image
        img_rm = self.resizeImage(img1)

        # Publish images
        try:
            pim1 = self.bridge.cv2_to_imgmsg(img_rm, encoding=self.IME_B)
            pim1.header = self.M_CAMERA_INFO.header

            pim2 = self.bridge.cv2_to_imgmsg(img2, encoding=self.IME_B)
            pim2.header = self.M_CAMERA_INFO.header

            # Publish image bands topics
            self.pub_1_.publish(pim1)
            self.pub_1_i.publish(self.M_CAMERA_INFO)
            self.pub_2_.publish(pim2)
            self.pub_2_i.publish(self.K_CAMERA_INFO)
        except CvBridgeError as e:
            print(e)

    def resizeImage(self, img):
        # Bicubic Interpolation for reseizing
        rimg = cv2.resize(img, (960, 764), interpolation=cv2.INTER_CUBIC)
        # Crop images to fit with depth images 960x540
        return rimg[112:rimg.shape[0]-112, :]

    def stop(self):
        self.sub_1_.unregister()
        self.sub_1_i.unregister()
        self.sub_2_.unregister()
        self.sub_2_i.unregister()
        self.pub_1_.unregister()
        self.pub_1_i.unregister()
        self.pub_2_.unregister()
        self.pub_2_i.unregister()
        rospy.signal_shutdown("Exit")
        rospy.loginfo("Synchronizer node (python) shutted down")


def main(args):
    rospy.init_node("synchronizer", anonymous=True)
    synchronizer = Synchronizer()
    try:
        rospy.spin()
    except KeyboardInterrupt:
        pass
    synchronizer.stop()


if __name__ == "__main__":
    main(sys.argv)
